examples/add/train_add_distill_sd_wds.py

```diff
@@ -0,0 +1,1673 @@
+                latents = []
+                for i in range(0, pixel_values.shape[0], 32):
+                    latents.append(vae.encode(pixel_values[i : i + 32]).latent_dist.sample())
+                latents = torch.cat(latents, dim=0)
+
+                latents = latents * vae.config.scaling_factor
+                latents = latents.to(weight_dtype)
+                bsz = latents.shape[0]
+
+                # 2. Sample random student timesteps s uniformly in `student_timestep_schedule` and sample random
+                # teacher timesteps t uniformly in [0, ..., noise_scheduler.config.num_train_timesteps - 1].
+                student_index = torch.randint(0, student_distillation_steps, (bsz,), device=latents.device).long()
+                student_timesteps = student_timestep_schedule[student_index]
+                teacher_timesteps = torch.randint(
+                    0, noise_scheduler.config.num_train_timesteps, (bsz,), device=latents.device
+                ).long()
+
+                # 3. Sample noise and add it to the latents according to the noise magnitude at each student timestep
+                # (that is, run the forward process on the student model)
+                student_noise = torch.randn_like(latents)
+                noisy_student_input = noise_scheduler.add_noise(latents, student_noise, student_timesteps)
+
+                # 4. Prepare prompt embeds and unet_added_conditions
+                prompt_embeds = encoded_text.pop("prompt_embeds")
+
+                # 5. Get the student model predicted original sample `student_x_0`.
+                student_noise_pred = unet(
+                    noisy_student_input,
+                    student_timesteps,
+                    encoder_hidden_states=prompt_embeds.float(),
+                ).sample
+                student_x_0 = denoiser(student_noise_pred, student_timesteps, noisy_student_input)
+
+                # 6. Sample noise and add it to the student's predicted original sample according to the noise
+                # magnitude at each teacher timestep (that is, run the forward process on the teacher model, but
+                # using `student_x_0` instead of latents sampled from the prior).
+                teacher_noise = torch.randn_like(student_x_0)
+                noisy_teacher_input = noise_scheduler.add_noise(student_x_0, teacher_noise, teacher_timesteps)
+
+                # 7. Get teacher model predicted original sample `teacher_x_0`.
+                with torch.no_grad():
+                    with torch.autocast("cuda"):
+                        teacher_noise_pred = teacher_unet(
+                            noisy_teacher_input.detach(),
+                            teacher_timesteps,
+                            encoder_hidden_states=prompt_embeds.float(),
+                        ).sample
+                        teacher_x_0 = denoiser(teacher_noise_pred, teacher_timesteps, noisy_teacher_input)
```

Comment:
As per the `sds` objective, I believe we should do a CFG step here.

Reply:
Makes sense. I guess probably in general even if we're not specifically using the SDS training weight function $`c(t) = \frac{\alpha_t}{2\sigma_t}w(t)`$ we still want to use the teacher CFG prediction as the distillation target?

(That brings up the question of how to choose the guidance weight $`\omega`$; I think probably sampling $`\omega \sim U([\omega_{min}, \omega_{max}])`$ like LCM does would probably be fine.)

Reply:
I would say there's definitely room for experimentation in how we choose the guidance weight, especially in light of https://github.com/huggingface/diffusers/pull/6303#discussion_r1436638310.

Reply:
Yes whether or not we are doing `SDS` we should do CFG step. Otherwise the samples generated by distilled model might be blurry.